From 48ec3f8d883e89650aec0f0130cdaa2eb945da5b Mon Sep 17 00:00:00 2001
From: Frank Neuber <fn@kernelport.com>
Date: Sun, 4 May 2014 16:46:07 +0200
Subject: [PATCH] * Fix for changed data structure in
 /usr/include/PCSC/reader.h   [Thank you to Bjarne Pohlers
 <debian@pohlers.org>]

---
 cjeca32/CCIDReader.cpp | 4 ++--
 cjeca32/stdafx.h       | 7 ++++++-
 doc/verifypin_ascii.c  | 4 ++--
 doc/verifypin_fpin2.c  | 4 ++--
 4 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/cjeca32/CCIDReader.cpp b/cjeca32/CCIDReader.cpp
index 1ab4135..c574868 100644
--- a/cjeca32/CCIDReader.cpp
+++ b/cjeca32/CCIDReader.cpp
@@ -3115,14 +3115,14 @@ RSCT_IFD_RESULT CCCIDReader::IfdVendor(uint32_t IoCtrlCode,uint8_t *Input,uint32
 	case CJPCSC_VEN_IOCTRL_VERIFY_PIN_DIRECT:
 		if(InputLength<sizeof(PIN_VERIFY_STRUCTURE))
          return STATUS_INVALID_BUFFER_SIZE;
-		if(InputLength!=sizeof(PIN_VERIFY_STRUCTURE)-sizeof(uint8_t)+((PIN_VERIFY_STRUCTURE *)Input)->ulDataLength)
+		if(InputLength!=sizeof(PIN_VERIFY_STRUCTURE)-sizeof(PIN_VERIFY_STRUCTURE::abData)+((PIN_VERIFY_STRUCTURE *)Input)->ulDataLength)
          return STATUS_INVALID_BUFFER_SIZE;
 		return IfdVerifyPinDirect((PIN_VERIFY_STRUCTURE *)Input,Output,OutputLength);
 
 	case CJPCSC_VEN_IOCTRL_MODIFY_PIN_DIRECT:
 		if(InputLength<sizeof(PIN_MODIFY_STRUCTURE))
          return STATUS_INVALID_BUFFER_SIZE;
-		if(InputLength!=sizeof(PIN_MODIFY_STRUCTURE)-sizeof(uint8_t)+((PIN_MODIFY_STRUCTURE *)Input)->ulDataLength)
+		if(InputLength!=sizeof(PIN_MODIFY_STRUCTURE)-sizeof(PIN_MODIFY_STRUCTURE::abData)+((PIN_MODIFY_STRUCTURE *)Input)->ulDataLength)
          return STATUS_INVALID_BUFFER_SIZE;
 		return IfdModifyPinDirect((PIN_MODIFY_STRUCTURE *)Input,Output,OutputLength);
   case CJPCSC_VEN_IOCTRL_SET_NORM:
diff --git a/cjeca32/stdafx.h b/cjeca32/stdafx.h
index 7149946..f90f3a9 100644
--- a/cjeca32/stdafx.h
+++ b/cjeca32/stdafx.h
@@ -56,7 +56,12 @@
 #include "ntstatus.h"
 
 #include "cjeca32.h"
-#include "PCSC10.h"
+#ifdef OS_LINUX
+  #include <PCSC/reader.h>
+#else
+  #include "PCSC10.h"
+#endif
+
 #include "Debug.h"
 #include "RSCTCriticalSection.h"
 #include "BaseCommunication.h"
diff --git a/doc/verifypin_ascii.c b/doc/verifypin_ascii.c
index c41d97b..d946c17 100644
--- a/doc/verifypin_ascii.c
+++ b/doc/verifypin_ascii.c
@@ -330,7 +330,7 @@ int main(int argc, char *argv[])
 	pin_verify -> abData[offset++] = 0x20;	/* '\0' */
 	pin_verify -> ulDataLength = htonl(offset);	/* APDU size */
 
-	length = sizeof(PIN_VERIFY_STRUCTURE) + offset -1;	/* -1 because PIN_VERIFY_STRUCTURE contains the first byte of abData[] */
+	length = sizeof(PIN_VERIFY_STRUCTURE) + offset - sizeof(PIN_VERIFY_STRUCTURE::abData);	/* because PIN_VERIFY_STRUCTURE may contain the first byte of abData[] */
 
 	printf(" command:");
 	for (i=0; i<length; i++)
@@ -463,7 +463,7 @@ int main(int argc, char *argv[])
 	pin_modify -> abData[offset++] = 0x30;	/* '0' */
 	pin_modify -> ulDataLength = HOST_TO_CCID_32(offset);	/* APDU size */
 
-	length = sizeof(PIN_MODIFY_STRUCTURE) + offset -1;	/* -1 because PIN_MODIFY_STRUCTURE contains the first byte of abData[] */
+	length = sizeof(PIN_MODIFY_STRUCTURE) + offset -sizeof(PIN_MODIFY_STRUCTURE::abData);	/* because PIN_MODIFY_STRUCTURE may contain the first byte of abData[] */;
 
 	printf(" command:");
 	for (i=0; i<length; i++)
diff --git a/doc/verifypin_fpin2.c b/doc/verifypin_fpin2.c
index 4ee4362..f236fdc 100644
--- a/doc/verifypin_fpin2.c
+++ b/doc/verifypin_fpin2.c
@@ -331,7 +331,7 @@ int main(int argc, char *argv[])
 	pin_verify -> abData[offset++] = 0xff;	/* '\0' */
 	pin_verify -> ulDataLength = htonl(offset);	/* APDU size */
 
-	length = sizeof(PIN_VERIFY_STRUCTURE) + offset -1;	/* -1 because PIN_VERIFY_STRUCTURE contains the first byte of abData[] */
+	length = sizeof(PIN_VERIFY_STRUCTURE) + offset - sizeof(PIN_VERIFY_STRUCTURE::abData);	/* because PIN_VERIFY_STRUCTURE may contain the first byte of abData[] */
 
 	printf(" command:");
 	for (i=0; i<length; i++)
@@ -464,7 +464,7 @@ int main(int argc, char *argv[])
 	pin_modify -> abData[offset++] = 0x30;	/* '0' */
 	pin_modify -> ulDataLength = HOST_TO_CCID_32(offset);	/* APDU size */
 
-	length = sizeof(PIN_MODIFY_STRUCTURE) + offset -1;	/* -1 because PIN_MODIFY_STRUCTURE contains the first byte of abData[] */
+	length = sizeof(PIN_MODIFY_STRUCTURE) + offset -sizeof(PIN_MODIFY_STRUCTURE::abData);	/* because PIN_MODIFY_STRUCTURE may contain the first byte of abData[] */
 
 	printf(" command:");
 	for (i=0; i<length; i++)
-- 
1.9.1

